IMOD/manpages/mrc2tif.man
.na
.nh
.TH mrc2tif 1 2.00 IMOD
.SH NAME
mrc2tif \- Create TIFF, JPEG, or PNG files from an MRC image file.
.SH SYNOPSIS
mrc2tif  [options]  mrc_input_file  tiff_file_root_name
.SH DESCRIPTION
mrc2tif usually creates a series of TIFF files with the prefix \fItiff_file_root_name\fR
and with the suffix .nnn.tif, where nnn is the z number.  Byte, signed and
unsigned 16-bit integer, float, 
and RGB color files (modes 0, 1, 6, 2, 16) can be converted.  The input
file need not be an MRC file; it can be any kind of file readable by
other IMOD programs, plus
the image file types readable by 3dmod using Qt libraries, of which the most
useful are PNG, JPEG, and BMP.
There are
options to create JPEG or PNG output files instead of TIFF.  The data will be
written to a TIFF file in the same mode as in the input file by default, but
there are options that allow the integer and float modes to be scaled to
bytes; the same options allow scaling of byte and RGB values.  If the input
file has only one section, then the output file will simply be the 
\fItiff_file_root_name\fR with no suffix added.  
.P
When the output mode is integer or float, the TIFF file will contain the
minimum and maximum value for the data in the file.  When a TIFF stack is
created, each
image will have the same overall minimum and maximum, rather than values
specific to that image.  This way, 3dmod will be able to read
in the data with proper scaling.  However, Photoshop pays no attention to
these values when displaying integer data, so when the integer data have a
relatively small range, the image will initially appear black or very dark.
The simplest solution for this problem is to apply the scaling options
described next.
.P
When either the \fB-S\fR, the \fB-C\fR, or the
.B -a
option is entered, integer and floating point data will be scaled to bytes and
any values outside the range of 0 to 255 will be truncated.  If the input data
is already in bytes (either byte data or RGB data), these options will cause
the data to be scaled, and values outside the range truncated.  To have the
range of the input data mapped into the full range of byte data, just enter
"-C 0,255".  To have a subset of the input range mapped to 0 to 255, either
enter the limits of the subset of the range directly with the 
.B -S
option, or adjust the contrast in 3dmod(1) to give the desired expansion of
the dynamic range for that subset of the image data, and enter the contrast
settings withs the 
.B -C
option.
.P
Output to JPEG or PNG files is selected by entering the
.B -j
or the
.B -p
option.  There are some restrictions in these cases.  An output stack cannot
be created, only a series of single-image files.  The type of compression
cannot be specified, but the quality or degree of compression can.  Tiling
cannot be specified.  Only byte and RGB output is available.  Integer and
floating point files are converted to byte based either on the minimum and maximum
densities in the input file or on the values entered 
with \fB-S\fR, \fB-C\fR, or \fB-a\fR.
In addition to the fact that JPEG images cannot be bigger than 65535 pixels in
X or Y, problems may be encountered with images much bigger than 1 gigapixel.
.P
Since IMOD is distributed with version 4 of the TIFF library, this program
will write files bigger than 4 GB in the BigTIFF format, and continue to write
smaller files in classical TIFF format.  
.P
When images are written to TIFF files with ZIP or LZW compression, the
program will use compress the data in parallel threads, with the number
of threads depending on the image size.  However, if the image is
larger than a certain threshold, it is processed in chunks instead of
read  in all at once, and the parallel compression routine cannot be
used.  The default threshold is 400 MB.  To ensure parallel compression,
you can either raise this threshold with the
.B -t
option or override the default behavior with "-O 1".  Note that with
parallel compression, memory usage will be about 2.2 times the size of the image.
.SH OPTIONS
.TP
.B -s
Create a single output file, a TIFF stack with all of
the sections in the input file.
.TP
.B -j
Produce JPEG files instead of TIFF files.  See above for the restrictions with
this kind of output.
.TP
.B -p
Produce PNG files instead of TIFF files.  See above for the restrictions.
.TP
.B -c \fIvalue\fR
Compress the data in TIFF files with the type of compression indicated.  The
most useful compression types are LZW (value 5), Zip or Deflate (value 8), and
JPEG (value 7).  For these types, the value can be given as a text string:
"lzw", "zip", or "jpeg".  The other allowed values are 32773 for
Packbits and 32946 for a possibly different Deflate.  These are
available on all operating systems when
running with the TIFF libraries distributed with IMOD.  
Also note that JPEG compression is not available with 16-bit data.
.TP
.B -q \fIvalue\fR
Set the quality for JPEG or ZIP compression (ZIP compression is used in PNG
output).  For JPEG, the quality ranges from
0 to 100 with a default of 75; higher numbers preserve more information but
with less compression.  For ZIP/PNG, which is a lossless compression, the quality
factor ranges from 1 to 9 with a default of 6 and specifies a tradeoff between
speed and amount of compression.  Higher numbers giving somewhat more
compression for more computational time.  Lower numbers give somewhat less
compression with potentially much less computational time.
.TP
.B -S \fImin,min\fR
Specify an initial intensity scaling that will scale "min" to 0 and "max" to
255, just as is achieved with the 
.B -s
options to 3dmod(1) and mrcbyte(1).  
.TP
.B -C \fIblack,white\fR
Specify intensity scaling to match the scaling achieved with the given
black and white contrast settings in 3dmod(1), with the -c option to
mrcbyte(1), or the "-contrast" option to newstack.  If both
.B -S
and
.B -C
are entered, the final scaling will be the product of these two scalings.
If
.B -S
is entered alone, the black/white values are 0/255 (no scaling for contrast).
If
.B -C
is entered alone, the initial scaling will be from the minimum in the input
file to 0, and from the maximum in the file to 255.  If you have integer data
and you want to convert it to bytes, you will generally want to apply some
scaling with black/white values to preserve the dynamic range of the data of
interest.
.TP
.B -a \fImean,SD\fR
Scale data for conversion to bytes so that it has the given mean and
standard deviation.  Enter 0 for either value to use the default
for auto-contrasting in 3dmod(1), which is a mean of 150 and an SD of 40.  This option
cannot be entered with \fB-S\fR or \fB-C\fR, since it supercedes the
scaling specified by them.
.TP
.B -z \fImin,max\fR
 Starting and ending Z slices in file to output, numbered from 0.
.TP
.B -i \fIvalue\fR
Initial file number.  The default is the number of the first Z slice output,
where Z is numbered from 0.
.TP
.B -r \fIvalue\fR
Set the resolution setting of the output file in dots per inch to the given
value.
.TP
.B -P
Use the pixel spacing in the MRC file header to set the resolution in
the output file.  The resolution will be set in dots per cm in a TIFF file and will
typically be a very large number, but this value allows the pixel spacing to
be recovered if the file is read back into IMOD programs or converted
back with Tif2mrc(1).
.TP
.B -T \fIxsize,ysize\fR or \fIsize\fR
Organize output data in tiles instead of strips, which can allow faster access
to subareas of very large images.  Enter one number for the tile size in both
dimensions, or two numbers to specify a different size in X and Y.  The TIFF
library requires that all tiles be the same size, which means that the image
file will be padded so that tiles on the bottom and right sides have
the same sizes as the rest.  The program will pick an actual tile size near
the entered number that is a multiple of 16 and that minimizes the amount of
padding required.
.TP
.B -t \fIvalue\fR
Set the criterion for processing each image in strips to the given value in
megabytes (fractional values are allowed).  Whole images will be 
read in, processed, and saved as a unit if they are less than this criterion
(default 400 MB).  Above the criterion, images will be read, processed, and
saved in strips (unless "-O 1" is entered), with much lower memory requirements.
.TP
.B -O \fIvalue\fR
Override the default behavior on whether to compress data in parallel.
Specifically, a value of 0 will force data not to be compressed in
parallel if it ordinarily would be.  A value of 1 will force large
images to be compressed in parallel if they would ordinarily be read in
chunks instead.
.TP
.B -o
Use old IMOD code instead of libtiff for writing the files (for testing
purposes).
.SH AUTHORS
Jim Kremer, David Mastronarde
.SH SEE ALSO
clip, newstack
.SH BUGS
Complex and long integer data modes are not supported.

Email bug reports to mast@colorado.edu.


IMOD/History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10/8/23: Fixed crash in Ctfphaseflip with X size of 4092 (i.e., K3 camera) and
tilt angle of exactly 0; fixed strip width at exactly 0 tilt with a maximum
width entered (In 4.11)

9/22/23: In Etomo - Fidual Model, fixed the "Use Adjusted Track Com File"
button, which was not functioning (In 4.11).(Bug# 2458)

3/14/23: Fixed Batchruntomo problem running Splitcorrection with Python 3
(In 4.11).

2/3/23: IMOD_4-11-24
2/3/23: Fixed Autofidseed problem with Python 3 (In 4.11).

1/10/22: IMOD_4-11-23
1/10/22: Fixed Sampletilt when the dataset name contains "slice" (In 4.11).

1/8/22: IMOD_4-11-22
12/22/22: Fixed a problem where most of the interfaces are unable to handle a
missing file name style (In 4.11).(Bug# 2427)

12/21/22: Fixed initialization of the back projection checkbox in batch
interface (in 4.11).(Bug# 2417)

12/7/22: Fixed Tiltalign crash from array that was too small when there are
many views and few points (In 4.11, already fixed in 4.12).

11/1/22: IMOD_4-11-21
11/1/22: Fixed failure in Copytomocoms from not converting some directive
values to float (In 4.11).

9/27/22: Add entry for autofidseed LowerTargetForClustered option.(Bug# 2418)

9/27/22: In Etomo - Batchruntomo Interface, fixed a problem with the
backprojection checkbox (In 4.11).(Bug# 2417)

9/7/22: Changes to framewatcher to watch for EER files (In 4.11).

9/2/22: IMOD_4-11-20
9/1/22: In Etomo - Batchruntomo Interface, fixed some problems with the back-
projection checkbox (In 4.11).(Bug# 2417)

8/25/22: Switched to Qt 5.15.12 for Linux builds to solve a problem with being
unable to resize or move windows after mouse-rotating in slicer (In 4.11).

8/3/22: Fixed 3dmod crash if model loaded during initial image load (In 4.11).

7/29/22: IMOD_4-11-19
7/29/22: Fixed no new line before MESSAGE: from Tilt if GPU not available
(In 4.11).

7/28/22: In Etomo - PEET, fixed a problem where the processchunks monitor did
not recognize any single-line messages from processchunks.out (In 4.11).
(Bug# 2321)

7/18/22: Changed defocus tolerance in ctfcorrection command file to 50 and
made Ctf3dsetup and Subtomosetup use the slab thickness for this tolerance, 
since it determines the width of regions near the edges corrected at just one
defocus (In 4.11).

7/12/22 In Etomo - Final Aligned Stack handling a config file with an
incorrect path (In 4.11).(Bug# 2404)

7/8/22 In Etomo - Tomogram Generation fixed the choosing of SIRT output
tomograms to load into 3dmod (In 4.11).(Bug# 2402)

7/7/22 In Etomo - Reconstruction Interface added missing image file name style
recovery (In 4.11).(Bug# 2403)

7/1/22: Re-enabled use of RGB data in Drawing Tools (In 4.11).

6/30/22: In Etomo - Batchruntomo Interface fixed a problem where a delivery
was done on every run (In 4.11).((Bug# 2405)

6/22/22: The minimum zero spacing in CTF correction is now defined as a float
and setting it will no longer generate an error in Etomo (In 4.11).(Bug# 2406)

6/18/22: IMOD_4-11-18
6/17/22: Fixed summing into power spectrum in Ctfplotter, which was off by a
pixel in the upper half of the FFT and biased defocus by 10-20 nm (In 4.11).

6/15/22: IMOD_4-11-17
6/14/22: Fixed wrong scaling of spherical aberration in Ctfphaseflip, which
made it correct effectively 20-50 nm higher defocus and get off above ~5 A
resolution (In 4.11).

6/11/22: Fixed Ctfplotter with pixel cropping to take tiles from correct
locations; it was sampling more from one side than the other, not adjusting
each spectrum properly for tile location, blurring the power spectrum, and
giving increasingly biased defocus estimates with increasing tilt (In 4.11).

6/10/22: In Etomo - Reconstruction interface, fixing a problem that prevented
a dataset created by batchruntomo from being loaded (In 4.11).(Bug# 2409)

6/8/22: Ctfplotter: fixed the crop pixel option when running
non-interactively; it was either not cropping or going into an infinite
loop. Fixed the starting point for autofitting to be near zero tilt 
when running non-interactively.  Fixed a memory leak (In 4.11).

6/6/22: Fixed Zap and Slicer snapshots having rubberband on edge with high-DPI
monitors (In 4.11).

6/2/22: Fixed crash in Ctfphaseflip with GPU from accessing strip coordinates
out of range (In 4.11).

5/30/22: IMOD_4-11-16
5/30/22: Made sure autocontrasting does not happen until the Zap window has
been drawn (In 4.11).

5/26/22: Made Batchruntomo detect and recover from an Etomo error in dataset
location when restarting from a step (In 4.11).

5/26/22: Fixed tile cache display in Zap window sometimes not being refreshed
after all the tiles are loaded (In 4.11).

5/25/22: Made Makepyramid work with a single-image file (In 4.11).

5/22/22: Changed installer for Linux/Mac to put link to python, if needed, in
a separate directory that is put on the back of the path (In 4.11).

5/20/22: Fixed skipping of global clip planes in model view (In 4.11).

5/20/22: Fixed sychronization of ghost and model view image display checkboxes
when toggling with hot keys (In 4.11).

5/20/22: Stopping limiting EER Z-summing to 99 in 3dmod (In 4.11).

5/6/22: Fixed a test for extension types in Splitblend, Splittilt, and
Splitcorrection (in 4.11).

4/26/22: Fixed direction of Step Up/Down buttons in Ctfplotter, broken 

IMOD/qttools/mrc2tif/mrc2tif.cpp
/*
 *  mrc2tif.cpp -- Convert mrc files to TIFF and other files.
 *
 *  Original author: James Kremer
 *  Revised by: David Mastronarde   email: mast@colorado.edu
 *
 *  Copyright (C) 1995-2016 by the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 *
 *  $Id$
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <qimage.h>
#include <qfile.h>
#include <qapplication.h>
#include <qstringlist.h>
#include "b3dtiff.h"
#include "b3dutil.h"
#include "parse_params.h"

static FILE *openEitherWay(ImodImageFile *iifile, char *iname, char *progname,
                           int oldcode);

static void usage(char *progname)
{
  printf("%s version %s \n", progname, VERSION_NAME);
  imodCopyright();
  printf("%s [options] <mrc file> <tiff name/root>\n\n", progname);
  printf(" Without -s, a series of tiff files will be created "
         "with the\n prefix [tiff root name] and with the suffix nnn.tif, "
         "where nnn is the z number. \n  Options:\n");
  printf("    -s         Stack all images in the mrc file into a single tiff"
         " file\n");
  printf("    -c val     Compress data; val can be lzw, zip, jpeg, or "
         "numbers defined\n\t\t in libtiff\n");
  printf("    -q #       Quality for jpeg compression (0-100) or for zip "
         "compression (1-9)\n");
  printf("    -S min,max Initial scaling limits for conversion to bytes\n");
  printf("    -C b,w     Contrast black/white values for conversion to "
         "bytes\n");
  printf("    -a mn,sd   Scale to mean and SD for conversion to bytes (0,0 for "
         "default)\n");
  printf("    -z min,max Starting and ending Z (from 0) to output\n");
  printf("    -i #       Initial file number (default is starting Z)\n");
  printf("    -j         Output jpeg file instead of tiff\n");
  printf("    -p         Output png file instead of tiff\n");
  printf("    -r #       Resolution setting in dots per inch\n");
  printf("    -P         Use pixel spacing in MRC header for resolution setting\n");
  printf("    -T nx[,ny] Output data in tiles of size nx by ny (nx by nx if "
         "ny omitted)\n");
    
  printf("    -O #       Override default for parallel tiff compression on whole"
         " image\n");
  printf("    -t #       Criterion image size in megabytes for processing "
         "file in strips\n");
  printf("    -o         Write file with old IMOD code instead of libtiff"
         "\n");
}

int main(int argc, char *argv[])
{
  FILE *fin;
  MrcHeader hdata;
  FILE *fpTiff = NULL;
  Islice slice;
  char *plugdir;
  QStringList strList;
  int xsize, ysize, zsize, psize, filenum, outPsize, numChunks, chunk;
  int lines, linesDone, version, linesPerChunk = 0, doChunks = 0;
  size_t xysize, allocSize, lineBytes;
  int i, j, slmode, z, iarg = 1, stack = 0, resolution = 0, initialNum = -1;
  int oldcode = 0, convert = 0, usePixel = 0;
  int compression = 1, black = 0, white = 255, zmin = -1, zmax = -1;
  int quality = -1;
  int autoContrast = 0;
  float autoMean = 150.;
  float autoSD = 40.;
  int makeJPG = 0;
  int makePNG = 0;
  int typeInd, digits = 3;
  bool makeQimage;
  b3dUInt32 ifdOffset = 0, dataOffset = 0;
  float chunkCriterion = 400.;
  float savecrit, dmin, dmax, smin =0., smax = 0.;
  int xTileSize = 0;
  int doParallel = -1, didParallel = 0;
  float val[3];
  float scale, offset, xscale, yscale, zscale, allMin, allMax, sliceMin, sliceMax;
  int numSample = 100000;
  float sample, imageMean, imageSD;
  unsigned char **linePtrs;
  char prefix[100];
  char format[14];
  IloadInfo li;
  ImodImageFile *iifile = iiNew();
  char *iname;
  unsigned char *buf, *qbuf, *qtmp;
  char *progname = imodProgName(argv[0]);

  /* 1 is none, 32773 is PACKBITS, 32946 is DEFLATE (non-Adobe) */
  int legalComp[] = {1, IICOMPRESSION_LZW, IICOMPRESSION_JPEG, IICOMPRESSION_ZIP, 32773,
                     32946};
  const char *extensions[3] = {"tif", "jpg", "png"};
  const char *typeNames[3] = {"TIFF", "JPEG", "PNG"};
  int numLegal = sizeof(legalComp) / sizeof(int);

  sprintf(prefix, "\nERROR: %s - ", progname);
  setExitPrefix(prefix);
  iiAddCheckFunction(iiQImageCheck);

  for (iarg = 1; iarg < argc - 1 ; iarg++){
    if (argv[iarg][0] == '-'){
      switch (argv[iarg][1]){
      case 's':
        stack = 1;
        break;

      case 'o':
        oldcode = 1;
        break;

      case 'c':
        iarg++;
        if (!strcmp(argv[iarg], "zip"))
          compression = IICOMPRESSION_ZIP;
        else if (!strcmp(argv[iarg], "jpeg"))
          compression = IICOMPRESSION_JPEG;
        else if (!strcmp(argv[iarg], "lzw"))
          compression = IICOMPRESSION_LZW;
        else
          compression = atoi(argv[iarg]);
        for (i = 0; i < numLegal; i++)
          if (compression == legalComp[i])
            break;
        if (i == numLegal)
          exitError("Compression value %d not allowed", compression);
        break;

      case 'r':
        resolution = -atoi(argv[++iarg]);
        break;

      case 'P':
        usePixel = 1;
        break;

      case 'p':
        makePNG = 1;
        break;

      case 'j':
        makeJPG = 1;
        break;

      case 'S':
        sscanf(argv[++iarg], "%f%*c%f", &smin, &smax);
        convert = 1;
        break;

      case 'C':
        sscanf(argv[++iarg], "%d%*c%d", &black, &white);
        convert = 1;
        break;

      case 'a':
        sscanf(argv[++iarg], "%f%*c%f", &scale, &offset);
        autoContrast = 1;
        if (scale)
          autoMean = scale;
        if (offset)
          autoSD = offset;
        break;

      case 'z':
        sscanf(argv[++iarg], "%d%*c%d", &zmin, &zmax);
        break;

      case 'i':
        initialNum = atoi(argv[++iarg]);
        break;

      case 'q':
        quality = atoi(argv[++iarg]);
        break;

      case 'T':
        sscanf(argv[++iarg], "%d%*c%d", &xTileSize, &linesPerChunk);
        if (!linesPerChunk)
          linesPerChunk = xTileSize;
        doChunks = 1;
        break;

      case 'O':
        doParallel = atoi(argv[++iarg]);
        break;

      case 't':
        chunkCriterion = atof(argv[++iarg]);
        break;

      case 'h':
        usage(progname);
        exit(1);

      default:
        break;
      }
      
    }
    else
      break;
       
  }

  if (oldcode && compression != 1)
    exitError("Compression not available with old writing code");
  if (oldcode && (resolution != 0 || usePixel))
    exitError("Resolution setting is not available with old writing code");
  if (oldcode && xTileSize)
    exitError("Tiling not available with old writing code");
  if (oldcode && (makeJPG || makePNG))
    exitError("JPEG and PNG output not available with old writing code");
  if (resolution != 0 && usePixel)
    exitError("You cannot enter both -r and -P");
  makeQimage = makeJPG || makePNG;
  typeInd = makeJPG + 2 * makePNG;
  if (makeQimage && (stack || doChunks || compression != 1 || 
                               xTileSize))
    exitError("You cannot enter -s, -c, or -T with JPEG and PNG output");
  if (makePNG && makeJPG)
    exitError("You cannot enter both -j and -p");
  if ((compression == 8 || makePNG) && (quality < -1 || quality == 0 || quality > 9))
    exitError("Quality for ZIP compression or PNG output must be between 1 and 9");

  if (convert && autoContrast)
    exitError("You cannot enter -C or -S for scaling with -a for auto-contrasting");
  if (autoContrast)
    convert = 1;

  if (argc - iarg != 2){
    usage(progname);
    exit(1);
  }

  // Do some initial things for Qimage output
  if (makeQimage) {

    // Put plugin dir on the library path so image plugins can be found
    plugdir = getenv("IMOD_PLUGIN_DIR");
    if (plugdir)
      strList << plugdir;
    plugdir = getenv("IMOD_DIR");
    if (plugdir)
      strList << QString(plugdir) + QString("/lib/imodplug");
    for (i = 0; i < strList.count(); i++)
      if (QFile::exists(strList[i] + "/imageformats"))
        break;
    if (strList.count() && i < strList.count())
      QApplication::setLibraryPaths(strList);

    // Convert the quality for QImage use by inverting and multiplying by 10 
    // Verified that the ZIP default of 6 is equivalent to Qt default of 30
    if (makePNG && quality >= 0)
        quality = 10 * (9 - quality);
  }
    
  // Get input file and header
  if (NULL == (fin = iiFOpen(argv[iarg], "rb")))
    exitError("Couldn't open %s", argv[iarg]);

  if (mrc_head_read(fin, &hdata))
    exitError("Can't Read Input Header from %s", argv[iarg]);
  iarg++;
  if (usePixel) {
    mrc_get_scale(&hdata, &xscale, &yscale, &zscale);
    resolution = 1.e8 / xscale;
  }

  // Adjust Z limits and file numbering
  if (zmin == -1 && zmax == -1) {
    zmin = 0;
    zmax = hdata.nz - 1;
  } else if (zmin < 0 || zmax >= hdata.nz || zmin > zmax)
    exitError("zmin,zmax values are reversed or out of the range 0 to %d\n",
              hdata.nz - 1);
  if (initialNum < 0)
    filenum = zmin;
  else
    filenum = initialNum;

  // Set up output file structure based on mode
  iifile->format = IIFORMAT_LUMINANCE;
  iifile->file = IIFILE_TIFF;
  iifile->type = IITYPE_UBYTE;
  iifile->amin = iifile->amax = 0;
  iifile->newFile = 1;

  switch(hdata.mode){
      
  case MRC_MODE_BYTE:
    psize = 1;
    break;
  case MRC_MODE_SHORT:
    psize = 2;
    iifile->type = IITYPE_SHORT;
    break;
  case MRC_MODE_USHORT:
    psize = 2;
    iifile->type = IITYPE_USHORT;
    break;
  case MRC_MODE_RGB:
    psize = 3;
    iifile->format = IIFORMAT_RGB;
    break;
  case MRC_MODE_FLOAT:
    psize = 4;
    iifile->type = IITYPE_FLOAT;
    break;
  default:
    exitError("Data mode %d not supported.", hdata.mode);
    break;
  }

  // Must convert to bytes if it is not color for PNG/JPEG output
  if (makeQimage && hdata.mode != MRC_MODE_BYTE && hdata.mode != MRC_MODE_RGB)
    convert = 1;
  xsize = hdata.nx;
  ysize = hdata.ny;
  zsize = hdata.nz;
  xysize = (size_t)xsize * (size_t)ysize;
  allocSize = xysize * psize;
  iifile->nx = xsize;
  iifile->ny = ysize;
  iifile->nz = 1;
  mrc_init_li(&li, NULL);
  li.xmin = 0;
  li.xmax = xsize - 1;
  li.ymin = 0;
  li.ymax = ysize - 1;
  dmin = hdata.amin;
  dmax = hdata.amax;
  slmode = sliceModeIfReal(hdata.mode);
  outPsize = psize;
  if (convert) {
    if (!autoContrast)
      mrcContrastScaling(&hdata, smin, smax, black, white, MRC_RAMP_LIN, &scale, 
                         &offset);
    if (slmode > 0) {
      iifile->type = IITYPE_UBYTE;
      outPsize = 1;
    }
  }

  if (!makeQimage) {

    // Check version and size and set up chunks for TIFF
    version = tiffVersion(&j);
    /* printf("Version %d.%d\n", version,j); */
    if (version < 4) {
      savecrit =  4.292e9;
      if (!version || oldcode)
        savecrit =  2.146e9;
      if ((double)xysize * outPsize >  savecrit)
        exitError("The image is too large in X/Y to save with TIFF version"
                  " %d.%d", version, j);
      if (stack && (zmax + 1 - zmin) * xysize * (double)outPsize >  savecrit)
        exitError("The volume is too large to save in a stack with TIFF version"
                  " %d.%d", version, j);
    }
    if (version > 0 && !oldcode && (doParallel <= 0 || compression == 1) && !autoContrast
        && ((double)xsize * ysize) * psize > chunkCriterion * 1024. * 1024.)
      doChunks = 1;

  } else {

    // Check size for JPEG
    if (makeJPG && (xsize > 65535 || ysize > 65535))
      exitError("The input image is too large in X or Y for JPEG output");

    // Round lines up to 32-bit boundary and get the number of bytes per output line
    // and fix allocated size if necessary, get line buffer
    lineBytes = 4 * ((xsize * outPsize + 3) / 4);
    if (allocSize < lineBytes * ysize)
      allocSize = lineBytes * ysize;
    qtmp = (unsigned char *)malloc(lineBytes);
    if (!qtmp)
      exitError("Failed to allocate memory for line buffer");
  }

  iname = (char *)malloc(strlen(argv[iarg]) + 20);
  if (!iname)
    exitError("Failed to allocate memory for name");

  /* Open new file for stack */
  if (stack) {
    sprintf(iname, "%s", argv[iarg]);
    iifile->nz = zmax + 1 - zmin;
    fpTiff = openEitherWay(iifile, iname, progname, oldcode);
  }    

  // Set up format for numbered filenames
  if (zmax + 1 - zmin >= 10000)
    digits = 5;
  else if (zmax + 1 - zmin >= 1000)
    digits = 4;
  sprintf(format, "%%s.%%0%dd.%%s", digits);

  allMin = 1.e30;
  allMax = -1.e30;

  printf("Writing %s images. ", typeNames[typeInd]);
  for (z = zmin; z <= zmax; z++){
    int tiferr;
    printf(".");
    fflush(stdout);

    /* Get filename and open new file for non-stack */
    if (!stack) {
      sprintf(iname, format, argv[iarg], filenum++, extensions[typeInd]);
      if (zsize == 1)
        sprintf(iname, "%s", argv[iarg]);
      if (!makeQimage) {
        if (z > zmin)
          free(iifile->filename);
        fpTiff = openEitherWay(iifile, iname, progname, oldcode);
        ifdOffset = 0;
        dataOffset = 0;
      }
    }

    // Prepare to get min/max for a slice, but send file min/max to slice initially
    sliceMin = 1.e30;
    sliceMax = -1.e30;
    iifile->amin = dmin;
    iifile->amax = dmax;
    
    /* If doing chunks, set up chunk loop */
    numChunks = 1;
    if (doChunks) {
      tiffWriteSetup(iifile, compression, 0, resolution, quality, 
                     &linesPerChunk, &numChunks, &xTileSize);
      if (xTileSize > xsize + 16 || linesPerChunk > ysize + 16)
        exitError("Entered tile size was too large");
      allocSize = xsize * linesPerChunk * psize;
      linesDone = 0;
    }
    
    for (chunk = 0; chunk < numChunks; chunk++) {

      /* Allocate memory first time or every time */
      if ((!chunk && z == zmin) || (convert && slmode > 0))
        qbuf = buf = (unsigned char *)malloc(allocSize);
      if (!buf)
        exitError("Failed to allocate memory for slice");

      lines = ysize;
      if (doChunks) {
        lines = B3DMIN(linesPerChunk, ysize - linesDone);
        li.ymin = ysize - (linesDone + lines);
        li.ymax = li.ymin + lines - 1;
        linesDone += lines;
      }

      if (mrcReadZ(&hdata, &li, buf, z)) {
        perror("mrc2tif ");
        exitError("Reading section %d", z);
      }
      sliceInit(&slice, xsize, lines, hdata.mode, buf);

      if (autoContrast) {
        linePtrs = makeLinePointers(slice.data.b, xsize, ysize, psize);
        if (!linePtrs)
          exitError("Allocating line pointers for autocontrasting");
        sample = B3DMIN(xsize * ysize, numSample) / (float)(xsize * ysize);
        i = sampleMeanSD(linePtrs, typeForSampleMean(hdata.mode), xsize, ysize, sample, 
                         0, 0, xsize, ysize, &imageMean, &imageSD);
        if (i)
          exitError("Error %d calling sampleMeanSD for autocontrasting", i);
        scale = autoSD / imageSD;
        offset = autoMean - scale * imageMean;
      }
      
      /* Scale the slice and convert it to bytes */
      if (convert) {
        for (j = 0; j < lines; j++) {
          for (i = 0; i < xsize; i++) {
            sliceGetVal(&slice, i, j, val);
            val[0] = val[0] * scale + offset;
            val[0] = B3DMAX(0., B3DMIN(255., val[0]));
            if (psize == 3) {
              val[1] = val[1] * scale + offset;
              val[1] = B3DMAX(0., B3DMIN(255., val[1]));
              val[2] = val[2] * scale + offset;
              val[2] = B3DMAX(0., B3DMIN(255., val[2]));
            }
            slicePutVal(&slice, i, j, val);
          }
        }
        if (slmode > 0 && sliceNewMode(&slice, SLICE_MODE_BYTE))
          exitError("Converting slice %d to bytes", z);
        qbuf = buf = slice.data.b;
        if (!buf)
          exitError("Failed to allocate memory for slice");
        if (makeQimage && lineBytes != xsize) {
          qbuf = (unsigned char *)malloc(lineBytes * ysize);
          if (!qbuf)
            exitError("Failed to allocate memory for unpacking image");
        }
      }
    
      // Get min/max for all image types, even bytes
      if (!makeQimage) {
        sliceMMM(&slice);
        sliceMin = B3DMIN(sliceMin, slice.min);
        sliceMax = B3DMAX(sliceMax, slice.max);
        allMin = B3DMIN(allMin, slice.min);
        allMax = B3DMAX(allMax, slice.max);
        if (!doChunks) {
          iifile->amin = sliceMin;
          iifile->amax = sliceMax;
        }
      }

      if (makeQimage) {

        // For Qimage, Spread out the image to give the aligned lines needed
        if (lineBytes != xsize * outPsize) {
          for (j = ysize - 1; j >= 0; j--)
            for (i = xsize * outPsize - 1; i >= 0; i--)
              qbuf[i + lineBytes * j] = buf[i + j * (size_t)xsize * outPsize];
        }

        // Invert the image
        for (j = 0; j < ysize / 2; j++) {
          memcpy(qtmp, &qbuf[lineBytes * j], lineBytes);
          memcpy(&qbuf[lineBytes * j], &qbuf[lineBytes * (ysize - 1 - j)], lineBytes);
          memcpy(&qbuf[lineBytes * (ysize - 1 - j)], qtmp, lineBytes);
        }

        // Create the QImage, set resolution if it exists, set color table for 8-bit
        QImage *qim = new QImage(qbuf, xsize, ysize, (int)lineBytes, outPsize > 1 ? 
                                 QImage::Format_RGB888 : QImage::Format_Indexed8);
        if (resolution) {
          qim->setDotsPerMeterX(resolution / (resolution > 0. ? 0.01 : -0.0254));
          qim->setDotsPerMeterY(resolution / (resolution > 0. ? 0.01 : -0.0254));
        }
        if (outPsize == 1)
          for (i = 0; i < 256; i++)
            qim->setColor(i, qRgb(i, i, i));
        tiferr = qim->save(iname, typeNames[typeInd], quality) ? 0 : 1;
        delete qim;
        if (qbuf != buf)
          free(qbuf);

        // Back to regular TIFF saving
      } else if (fpTiff)
        tiferr = tiff_write_image(fpTiff, xsize, ysize, hdata.mode, buf, 
                                  &ifdOffset, &dataOffset, dmin, dmax);
      else if (!doChunks && doParallel) 
        tiferr = tiffParallelWrite(iifile, buf, compression, 0, resolution,
                                   quality, &didParallel);
      else if (!doChunks) 
        tiferr = tiffWriteSection(iifile, buf, compression, 0, resolution,
                                  quality);
      else
        tiferr = tiffWriteStrip(iifile, chunk, buf);
      if (tiferr) {
        exitError("Error (%d) writing section %d to %s%s", tiferr, z, iname, 
                  didParallel ? "; you could use option -O 0 to try again without "
                  "parallelized compression" : "");
      }
      if (convert && slmode > 0)
        free(slice.data.b);
    }

    // Finish up the file; set the min/max for real now
    iifile->amin = sliceMin;
    iifile->amax = sliceMax;
    if (doChunks)
      tiffWriteFinish(iifile);
    
    if (!stack && !makeQimage) {
      if (fpTiff) {
        fclose(fpTiff);
        fpTiff = NULL;
      } else
        iiClose(iifile);
    }
  }

  printf("\r\n");

  // Finish up the stack; give the last image the full min/max
  iifile->amin = allMin;
  iifile->amax = allMax;
  if (stack) {
    if (fpTiff)
      fclose(fpTiff);
    else
      iiClose(iifile);
  }
  iiFClose(fin);
  if (xTileSize)
    printf("Actual tile size = %d x %d\n", xTileSize, linesPerChunk);
  exit(0);
}

static FILE *openEitherWay(ImodImageFile *iifile, char *iname, char *progname,
                           int oldcode)
{
  static int warned = 0;
  FILE *fpTiff = NULL;
  iifile->filename = strdup(iname);
  
  if (oldcode || tiffOpenNew(iifile)) {
    fpTiff = fopen(iname, "wb");
    if (!fpTiff){
      perror("mrc2tif system message");
      exitError("Opening %s", iname);
    } else if (!oldcode && !warned) {
      warned = 1;
      printf("\nWARNING: %s - Not writing with libtiff, compression not "
             "available", progname);
    }
  }
  return fpTiff;
}
